## Installing Base packages

In [ ]:
!uv add pandas matplotlib seaborn scikit-learn
# !pip install pandas matplotlib seaborn scikit-learn # use this line if you are not using uv

## Data Science Assignment-1
### Project Objective

The goal of this analysis is to perform a comprehensive Exploratory Data Analysis (EDA) on pharmaceutical sales data to identify buying patterns, seasonal trends, and correlations between different drug categories. This information is critical for inventory optimization and supply chain management in a retail pharmacy context.

### Dataset Selection
For this assignment, we will use the [Pharma sales data](https://www.kaggle.com/datasets/milanzdravkovic/pharma-sales-data?resource=download) by Milan Zdravković available in Keggle. Our focus will on the _salesdaily.csv_ dataset.This dataset contains daily sales records of various drug categories classified by the ATC (Anatomical Therapeutic Chemical) system.

**Why this Dataset?** --> Daily granularity allows for the identification of weekday vs. weekend patterns and specific date-related spikes that would be lost in weekly or monthly aggregations. 

#### Step-1 Data Loading and Base analysis

In this step we will load the dataset and perform base exploratory analysis

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from pathlib import Path

DATASET_PATH = Path.cwd()/"data"/"pharma_sales_data"/"salesdaily.csv"

try:
    df:pd.DataFrame = pd.read_csv(DATASET_PATH)
    print("Dataset loaded successfully!")
    print(f"Dataset shape: {df.shape} rows and {df.shape[1]} columns")
    print(f"Dataset columns: {df.columns.tolist()}")
except FileNotFoundError:
    print(f"Error: Dataset file not found at {DATASET_PATH}")
except Exception as e:
    print(f"An error occurred while loading the dataset: {e}")

# Display the first few rows of the dataset to understand its structure
df.head()

Dataset loaded successfully!
Dataset shape: (2106, 13) rows and 13 columns
Dataset columns: ['datum', 'M01AB', 'M01AE', 'N02BA', 'N02BE', 'N05B', 'N05C', 'R03', 'R06', 'Year', 'Month', 'Hour', 'Weekday Name']


,datum,M01AB,M01AE,N02BA,N02BE,N05B,N05C,R03,R06,Year,Month,Hour,Weekday Name
0,1/2/2014,0.0,3.67,3.4,32.40,7.0,0.0,0.0,2.0,2014,1,248,Thursday
1,1/3/2014,8.0,4.00,4.4,50.60,16.0,0.0,20.0,4.0,2014,1,276,Friday
2,1/4/2014,2.0,1.00,6.5,61.85,10.0,0.0,9.0,1.0,2014,1,276,Saturday
3,1/5/2014,4.0,3.00,7.0,41.10,8.0,0.0,3.0,0.0,2014,1,276,Sunday
4,1/6/2014,5.0,1.00,4.5,21.70,16.0,2.0,6.0,2.0,2014,1,276,Monday


#### Initial Observations:
The dataset consists of a datum column and several columns representing drug categories (_M01AB_, _N02BA_, etc.).There are auxiliary columns like _Year_, _Month_, and _Weekday Name_ that will be useful for time-series analysis. The _datum_ column appears to be a string/object type and will require conversion in the Data Cleaning step.

#### Step 2: Basic Data Exploration
In this step we will perform initial data analysis to understand its strctural characteristics. This will include
- Inspecting the first and last few rows to verify data consistency.
- Checking the dimensions (number of rows and columns) of the dataset
- Identifying the data types of each feature (Numeric vs. Categorical)
- Generating a statistical summary to understand the distribution of sales across different ATC categories.

In [5]:
# 1. Inspecting the first 5 rows
print("--- First 5 Rows ---")
display(df.head())

# 2. Checking the dimensions of the dataset
print(f"\nDataset Dimensions: {df.shape[0]} rows and {df.shape[1]} columns")

# 3. Checking data types and non-null counts
print("\n--- Data Information ---")
df.info()

# 4. Statistical summary of the numeric columns
print("\n--- Statistical Summary ---")
display(df.describe())

--- First 5 Rows ---


,datum,M01AB,M01AE,N02BA,N02BE,N05B,N05C,R03,R06,Year,Month,Hour,Weekday Name
0,1/2/2014,0.0,3.67,3.4,32.40,7.0,0.0,0.0,2.0,2014,1,248,Thursday
1,1/3/2014,8.0,4.00,4.4,50.60,16.0,0.0,20.0,4.0,2014,1,276,Friday
2,1/4/2014,2.0,1.00,6.5,61.85,10.0,0.0,9.0,1.0,2014,1,276,Saturday
3,1/5/2014,4.0,3.00,7.0,41.10,8.0,0.0,3.0,0.0,2014,1,276,Sunday
4,1/6/2014,5.0,1.00,4.5,21.70,16.0,2.0,6.0,2.0,2014,1,276,Monday



Dataset Dimensions: 2106 rows and 13 columns

--- Data Information ---
<class 'pandas.DataFrame'>
RangeIndex: 2106 entries, 0 to 2105
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   datum         2106 non-null   str    
 1   M01AB         2106 non-null   float64
 2   M01AE         2106 non-null   float64
 3   N02BA         2106 non-null   float64
 4   N02BE         2106 non-null   float64
 5   N05B          2106 non-null   float64
 6   N05C          2106 non-null   float64
 7   R03           2106 non-null   float64
 8   R06           2106 non-null   float64
 9   Year          2106 non-null   int64  
 10  Month         2106 non-null   int64  
 11  Hour          2106 non-null   int64  
 12  Weekday Name  2106 non-null   str    
dtypes: float64(8), int64(3), str(2)
memory usage: 214.0 KB

--- Statistical Summary ---


,M01AB,M01AE,N02BA,N02BE,N05B,N05C,R03,R06,Year,Month,Hour
count,2106.000000,2106.000000,2106.000000,2106.000000,2106.000000,2106.000000,2106.000000,2106.000000,2106.000000,2106.000000,2106.000000
mean,5.033683,3.895830,3.880441,29.917095,8.853627,0.593522,5.512262,2.900198,2016.401235,6.344255,275.945869
std,2.737579,2.133337,2.384010,15.590966,5.605605,1.092988,6.428736,2.415816,1.665060,3.386954,1.970547
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2014.000000,1.000000,190.000000
25%,3.000000,2.340000,2.000000,19.000000,5.000000,0.000000,1.000000,1.000000,2015.000000,3.000000,276.000000
50%,4.990000,3.670000,3.500000,26.900000,8.000000,0.000000,4.000000,2.000000,2016.000000,6.000000,276.000000
75%,6.670000,5.138000,5.200000,38.300000,12.000000,1.000000,8.000000,4.000000,2018.000000,9.000000,276.000000
max,17.340000,14.463000,16.000000,161.000000,54.833333,9.000000,45.000000,15.000000,2019.000000,12.000000,276.000000


#### Initial Findings
- Data Volume: Dataset contains 2106 records, providing robust sample for daily trend analysis/
- Data Cleanliness: From df.info() we can infer there is no significant null data , thus data integrety is maintained. But some additioanl data prep will be needed due to mismatched data types.
- Features: The columns represent specific drug classifications (ATC codes) such as M01AB, N02BA, N02BE, N05B, etc.
- Most sales columns are floating-point numbers (float64), which is expected for quantity data. However, the datum column is currently stored as an str and will need to be converted to datetime in the next step.
- Scale: The describe() output shows a significant difference in sales volume between categories; for instance, some categories have a much higher daily average than others, suggesting certain medications are "high-movers."

### Step 3: Data Cleaning and harmonization

In the initial exploration we found that the data does not contain any null records, but we need to refine the data structure for the time-series analysis.

- **Data Formatting**: The _datum_ column is currently a string, we must convert it to a datetime object so we can extract seasonal and monthly trend.
- **Redundancy Removal**: Since this is a daily sales dataset, the Hour column is redundant , and will be dropped.
- **Consistency Check**: We will ensure all sales columns are properly numeric and handle any hidden "ghost" markers.

In [6]:
# Converting datum to datetime format
df['datum'] = pd.to_datetime(df['datum'], errors='coerce')

# Removing redundant columns
columns_to_drop = ['Hour']

df = df.drop(columns=columns_to_drop)

# Ensure numeric columns are strictly float (handling hidden non-numeric strings)
# ATC categories are from column index 1 to 8
atc_columns = ['M01AB', 'M01AE', 'N02BA', 'N02BE', 'N05B', 'N05C', 'R03', 'R06']

for col in atc_columns:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Final check for any NaNs created by 'coerce'
print("Missing values after cleaning:")
print(df.isnull().sum())

# Display cleaned info
df.info()

Missing values after cleaning:
datum           0
M01AB           0
M01AE           0
N02BA           0
N02BE           0
N05B            0
N05C            0
R03             0
R06             0
Year            0
Month           0
Weekday Name    0
dtype: int64
<class 'pandas.DataFrame'>
RangeIndex: 2106 entries, 0 to 2105
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   datum         2106 non-null   datetime64[us]
 1   M01AB         2106 non-null   float64       
 2   M01AE         2106 non-null   float64       
 3   N02BA         2106 non-null   float64       
 4   N02BE         2106 non-null   float64       
 5   N05B          2106 non-null   float64       
 6   N05C          2106 non-null   float64       
 7   R03           2106 non-null   float64       
 8   R06           2106 non-null   float64       
 9   Year          2106 non-null   int64         
 10  Month         2106 non-null   int64    

#### Cleaning findings
- **Feature Transformation**: The _datum_ column has been successfully converted to __*datetime64[us]*__ enabling time-based filtering.
- **Dimensionaluty Reduction**: Removed hour column, reducing noise in the data.
- **Data Integrity**: No null values were generated during the numeric conversion, confirming that the sales columns do not contain hidden non-numeric markers.


In [8]:
# Saving Cleaned Dataset as a checkpoint before further analysis
CLEANED_DATASET_PATH = Path.cwd()/"data"/"pharma_sales_data"/"cleaned"/"salesdaily_cleaned.csv"

try:
    df.to_csv(CLEANED_DATASET_PATH, index=False)
    print(f"Cleaned dataset saved successfully at {CLEANED_DATASET_PATH}")
except Exception as e:
    print(f"An error occurred while saving the cleaned dataset: {e}")

Cleaned dataset saved successfully at d:\MBA_Assignment\Data_Science_Assignment\Sem2_DS_Assignment\data\pharma_sales_data\cleaned\salesdaily_cleaned.csv
